### PR TITLE
#49973 Added the convex notch shape option.

### PR DESCRIPTION
**Pull request by domske**
## Description

Added a convex option for the notch shape. (see pictures at the end of this message.)

## Related Issues

#49973 

## Tests

This option is not yet tested. Feel free to add this. That would be nice. (Or I will do that soon.)

## Checklist

- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I signed the [CLA].
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [ ] All existing and new tests are passing. (see Tests section above.)
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

- [X] No, no existing tests failed, so this is *not* a breaking change. (It's optional.)

## Screenshots

**Default notch shape (concave) without body extended.**

concave notch and not extended body

**Default notch shape (concave) with body extended.**

concave notch and extended body

**New optional notch shape (convex) without body extended.**

convex notch and not extended body

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter/lib/src/painting/notched_shapes.dart [hunk 1]
```diff
@@ -37,10 +37,15 @@ abstract class NotchedShape {
 ///
 ///  * [CircleBorder], a [ShapeBorder] that describes a circle.
 class CircularNotchedRectangle extends NotchedShape {
+  /// A convex notch shape. It's concave by default.
```
Comment (Piinks) [inline packages/flutter/lib/src/painting/notched_shapes.dart]: Nit: Dart docs specify the first sentence should be its own paragraph. https://dart.dev/guides/language/effective-dart/documentation#do-separate-the-first-sentence-of-a-doc-comment-into-its-own-paragraph

File: packages/flutter/lib/src/painting/notched_shapes.dart [hunk 2]
```diff
@@ -37,10 +37,15 @@ abstract class NotchedShape {
...
+  final bool convex;
+
   /// Creates a [CircularNotchedRectangle].
   ///
+  /// `convex` Inverts the default concave shape to a convex shape.
```
Comment (Piinks) [inline packages/flutter/lib/src/painting/notched_shapes.dart]: What does concave and convex mean in this context here? Is the convex option technically a notched shape?

**Review by Piinks — COMMENTED** [review]
Hi @domske, welcome :tada: and thank you for the contribution!
It looks like you will need to sign the cla and provide tests for this change. The analyzer is also currently failing.

**Comment by domske** [comment]
@Piinks Thanks for the review. Unfortunately I have currently no time for unit tests and CLA. Also I have currently problems to get Flutter repo run. (`sample_catalog depends on intl 0.16.0 / 0.16.1` etc.) I'm new to Dart and Flutter and that's very time consuming right now. Sorry.
Feel free to complete this RP or create a new one. The important source code snippets are also in the issue #49973.

File: packages/flutter/lib/src/painting/notched_shapes.dart (hunk 2 above)
Comment (domske) [reply packages/flutter/lib/src/painting/notched_shapes.dart]: This means that one notch shape goes inside (concave) and the other goes outside (convex). But we could also name it just "invert". Or create a new shape without options. What do you think? I thought to invert the path Y direction by an option will reduce repetition code. This is just a proposal. :)

File: packages/flutter/lib/src/painting/notched_shapes.dart (hunk 1 above)
Comment (domske) [reply packages/flutter/lib/src/painting/notched_shapes.dart]: A trailing line like that:

```dart
/// A convex notch shape. It's concave by default.
///
final bool convex;
```

? Ok.

**Pull request closed without merging**